Note : The following change was trimmed. 3rd party code update was dropped
from the patch (unneeded for our package)

From c8e56913ff414fe18d8ba5675819c7eaef4946cb Mon Sep 17 00:00:00 2001
From: Shawn Rutledge <shawn.rutledge@qt.io>
Date: Wed, 23 Sep 2026 22:23:16 +0200
Subject: [PATCH] Update md4c to 0.6.0
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Amends d377d1f3a921488525f6e75c850632f559707747 : Qt's definition of
GitHub dialect remains as it was, while md4c has recently added features
that we don't yet support. We need to verify that the new features are
not inadvertently turned on before Qt is ready to handle them, and deal
with the clashing feature flag (0x100000 is for FrontMatter in Qt,
footnotes in md4c). It seems for now that we will not be able to support
footnotes, both because of the clashing flag and because it would be
hard to change the text-editing components to support them; so a test is
added to verify that the markdown footnote syntax is NOT interpreted
specially. Likewise, a test verifies that a markdown "admonition" is
treated as an ordinary block quote for now (but this can be improved
later).

[ChangeLog][Third-Party Code] Updated md4c to version 0.6.0.

Fixes: QTBUG-148738
Fixes: QTBUG-150636
Fixes: QTBUG-150661
Task-number: QTBUG-149921
Pick-to: 6.12.0 6.8 6.5
Change-Id: Iffd17982fe0d606b0f5003e25ee85b2db6ad019a
Reviewed-by: Oliver Eftevaag <oliver.eftevaag@qt.io>
Reviewed-by: Kai Köhne <kai.koehne@qt.io>
(cherry picked from commit 032643f99e090e0304ef935c2ee11c109d55c8a1)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---

diff --git a/src/gui/text/qtextmarkdownimporter.cpp b/src/gui/text/qtextmarkdownimporter.cpp
index 3f57c700fa8..3ee8b4104de 100644
--- a/src/gui/text/qtextmarkdownimporter.cpp
+++ b/src/gui/text/qtextmarkdownimporter.cpp
@@ -51,8 +51,10 @@ static_assert(int(QTextMarkdownImporter::FeaturePermissiveAutoLinks) == MD_FLAG_
 static_assert(int(QTextMarkdownImporter::FeatureTasklists) == MD_FLAG_TASKLISTS);
 static_assert(int(QTextMarkdownImporter::FeatureNoHTML) == MD_FLAG_NOHTML);
 static_assert(int(QTextMarkdownImporter::DialectCommonMark) == MD_DIALECT_COMMONMARK);
+// We don't yet support all the features in MD_DIALECT_GITHUB
 static_assert(int(QTextMarkdownImporter::DialectGitHub) ==
-              (MD_DIALECT_GITHUB | MD_FLAG_UNDERLINE | QTextMarkdownImporter::FeatureFrontMatter));
+              (MD_FLAG_PERMISSIVEAUTOLINKS | MD_FLAG_TABLES | MD_FLAG_STRIKETHROUGH | MD_FLAG_TASKLISTS
+                | MD_FLAG_UNDERLINE | QTextMarkdownImporter::FeatureFrontMatter));
 
 // --------------------------------------------------------
 // MD4C callback function wrappers
@@ -167,7 +169,7 @@ void QTextMarkdownImporter::import(const QString &markdown)
 {
     MD_PARSER callbacks = {
         0, // abi_version
-        unsigned(m_features),
+        unsigned(m_features) & unsigned(0xFFEFFFFF), // exclude MD_FLAG_FOOTNOTES: unsupported in Qt
         &CbEnterBlock,
         &CbLeaveBlock,
         &CbEnterSpan,
diff --git a/src/gui/text/qtextmarkdownimporter_p.h b/src/gui/text/qtextmarkdownimporter_p.h
index 8b8f4ec9bb8..ba188a3b33e 100644
--- a/src/gui/text/qtextmarkdownimporter_p.h
+++ b/src/gui/text/qtextmarkdownimporter_p.h
@@ -47,6 +47,10 @@ public:
         FeatureTasklists =                0x0800,
         FeatureUnderline =                0x4000,
         FeatureFrontMatter =            0x100000, // Qt feature, not yet in MD4C
+        // NOTE: flags above are kept in sync with #define MD_FLAG_*; therefore,
+        // do NOT add any more Qt-specific flags (within the lower 32 bits).
+        // Because we have FeatureFrontMatter, we cannot support MD_FLAG_FOOTNOTES.
+
         // composite flags
         FeaturePermissiveAutoLinks = FeaturePermissiveMailAutoLinks
             | FeaturePermissiveURLAutoLinks | FeaturePermissiveWWWAutoLinks,
diff --git a/tests/auto/gui/text/qtextmarkdownimporter/data/admonitions.md b/tests/auto/gui/text/qtextmarkdownimporter/data/admonitions.md
new file mode 100644
index 00000000000..040d133fbd7
--- /dev/null
+++ b/tests/auto/gui/text/qtextmarkdownimporter/data/admonitions.md
@@ -0,0 +1,17 @@
+Admonitions, AKA Alerts:
+https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts
+
+> [!NOTE]
+> Useful information that users should know, even when skimming content.
+
+> [!TIP]
+> Helpful advice for doing things better or more easily.
+
+> [!IMPORTANT]
+> Key information users need to know to achieve their goal.
+
+> [!WARNING]
+> Urgent info that needs immediate user attention to avoid problems.
+
+> [!CAUTION]
+> Advises about risks or negative outcomes of certain actions.
diff --git a/tests/auto/gui/text/qtextmarkdownimporter/data/footnotes.md b/tests/auto/gui/text/qtextmarkdownimporter/data/footnotes.md
new file mode 100644
index 00000000000..91c3f84011a
--- /dev/null
+++ b/tests/auto/gui/text/qtextmarkdownimporter/data/footnotes.md
@@ -0,0 +1,13 @@
+From https://www.markdownguide.org/extended-syntax/#footnotes
+
+Here's a simple footnote,[^1] and here's a longer one.[^bignote]
+
+[^1]: This is the first footnote.
+
+[^bignote]: Here's one with multiple paragraphs and code.
+
+    Indent paragraphs to include them in the footnote.
+
+    `{ my code }`
+
+    Add as many paragraphs as you like.
diff --git a/tests/auto/gui/text/qtextmarkdownimporter/tst_qtextmarkdownimporter.cpp b/tests/auto/gui/text/qtextmarkdownimporter/tst_qtextmarkdownimporter.cpp
index 001f3ff00d7..0f0dd80a8da 100644
--- a/tests/auto/gui/text/qtextmarkdownimporter/tst_qtextmarkdownimporter.cpp
+++ b/tests/auto/gui/text/qtextmarkdownimporter/tst_qtextmarkdownimporter.cpp
@@ -46,6 +46,8 @@ private slots:
     void fencedCodeBlocks();
     void frontMatter_data();
     void frontMatter();
+    void admonitions();
+    void footnotes();
     void toRawText_data();
     void toRawText();
 
@@ -719,6 +721,90 @@ void tst_QTextMarkdownImporter::frontMatter()
     QCOMPARE(doc.metaInformation(QTextDocument::FrontMatter).size(), expectedFrontMatterSize);
 }
 
+void tst_QTextMarkdownImporter::admonitions() // QTBUG-149922
+{
+    // QTextDocument doesn't have a representation for admonitions yet.
+    // But test what happens when we parse a markdown file that has them.
+    // We expect them to be seen as block quotes; the syntax like "[!NOTE]"
+    // comes out as-is for now, whereas in the future it could trigger some CSS style.
+    QTextDocument doc;
+    QFile f(QFINDTESTDATA("data/admonitions.md"));
+    QVERIFY(f.open(QFile::ReadOnly));
+    doc.setMarkdown(f.readAll());
+
+#ifdef DEBUG_WRITE_HTML
+    {
+        QFile out("/tmp/admonitions.html");
+        out.open(QFile::WriteOnly);
+        out.write(doc.toHtml().toLatin1());
+        out.close();
+    }
+#endif
+
+    QTextFrame::iterator iterator = doc.rootFrame()->begin();
+    QTextFrame *currentFrame = iterator.currentFrame();
+    int blockQuoteCount = 0;
+    int plainBlockCount = 0;
+    while (!iterator.atEnd()) {
+        // There are no child frames
+        QCOMPARE(iterator.currentFrame(), currentFrame);
+        // Check BlockQuoteLevel
+        QTextBlock block = iterator.currentBlock();
+        const bool blockQuote = block.blockFormat().hasProperty(QTextFormat::BlockQuoteLevel);
+        if (blockQuote) {
+            QCOMPARE(block.blockFormat().intProperty(QTextFormat::BlockQuoteLevel), 1);
+            ++blockQuoteCount;
+        } else {
+            ++plainBlockCount;
+        }
+        qCDebug(lcTests) << (blockQuoteCount ? "quote" : "text") << block.text();
+        ++iterator;
+    }
+    QCOMPARE(blockQuoteCount, 5);
+    QCOMPARE(plainBlockCount, 1);
+}
+
+void tst_QTextMarkdownImporter::footnotes() // QTBUG-149923
+{
+    // QTextDocument doesn't have a representation for footnotes yet.
+    // But test what happens when we parse a markdown file that has them.
+    QTextDocument doc;
+    QFile f(QFINDTESTDATA("data/footnotes.md"));
+    QVERIFY(f.open(QFile::ReadOnly));
+    doc.setMarkdown(f.readAll());
+
+#ifdef DEBUG_WRITE_HTML
+    {
+        QFile out("/tmp/footnotes.html");
+        out.open(QFile::WriteOnly);
+        out.write(doc.toHtml().toLatin1());
+        out.close();
+    }
+#endif
+
+    QTextFrame::iterator iterator = doc.rootFrame()->begin();
+    QTextFrame *currentFrame = iterator.currentFrame();
+    int nonEmptyBlockCount = 0;
+    int firstFootNoteIndex = -1;
+    while (!iterator.atEnd()) {
+        // There are no child frames
+        QCOMPARE(iterator.currentFrame(), currentFrame);
+        QTextBlock block = iterator.currentBlock();
+        if (!block.text().isEmpty()) {
+            qCDebug(lcTests) << block.text();
+            // Without MD_FLAG_FOOTNOTES (which QTD cannot support right now),
+            // the special syntax is ignored, and the first footnote is an ordinary inline paragraph.
+            // If MD_FLAG_FOOTNOTES got turned on, md4c would emit it into the last block instead.
+            if (block.text().contains("This is the first footnote."))
+                firstFootNoteIndex = nonEmptyBlockCount;
+            ++nonEmptyBlockCount;
+        }
+        ++iterator;
+    }
+    QCOMPARE(nonEmptyBlockCount, 7);
+    QCOMPARE(firstFootNoteIndex, 2);
+}
+
 void tst_QTextMarkdownImporter::toRawText_data()
 {
     QTest::addColumn<QString>("input");
-- 
2.55.0

